### DISCUSSION TITLE
Environment management and UWSM

### DISCUSSION THREAD

**Discussion by Vladimir-csp** [labwc/labwc#1801] [Q&A]
Hi!
I'm refreshing support for labwc in my [session manarer](https://github.com/Vladimir-csp/uwsm) and noticed that at some point environment files for labwc migrated from `env` to `environment`, and they are no longer shell code.

Is there any difference between setting a var in native labwc `environment` file and having a var pre-set before labwc is started?

Are vars defined in native labwc `environment` files exported to activation environments by labwc when `LABWC_UPDATE_ACTIVATION_ENV` is truthy? On config reload? Are they cleaned up on exit?

I was going to add `export LABWC_UPDATE_ACTIVATION_ENV=0` to labwc plugin (because uwsm manages environment itself). But it also sourced `~/.config/labwc/env`, which is no longer a valid thing to do and has to be removed.

I'd like to find a non-conflicting and non-intrusive way to manage environment for labwc+uwsm. Since `environment` is not shell code, it can not be sourced on uwsm side. Variable names from `environment` can be manually added to `uwsm finalize` command in the `autostart`. It would be a hassle, but should work.

If labwc could define a variable containing names of all the variables it got in its `environment` files (plus any vars of interest), this task would be reduced to having something like `uwsm finalize $LABWC_NATIVE_VARNAMES`. Or if labwc manages those vars in activation environments itself, then special treatment on uwsm side would not be needed at all.

**Comment by ahesford** [comment]
As far as I know, the `environment` file was never a shell script. That's not the case with Openbox, which uses a session-manager wrapper to source the environment file and establish the environment before the actual executable runs.

I do not believe that there is a substantive difference between setting files via the `environment` file or setting them before running `labwc`, because the program will parse the file and set the environment before doing anything else. However, there may be some corner cases that I don't see.

During session initialization, we export a [limited set of variables](https://github.com/labwc/labwc/blob/master/src/config/session.c#L24-L31) to systemd and D-Bus activation environments, if possible. On exit, we tear those down as best we can. (With systemd, those variables can actually be unset; with D-Bus, the closest that we can get is setting those variables to empty strings.) I do not think we should assume further responsibility for setting more variables in these activation environments. We set what we do because 1) we directly control a few of them, and 2) the cursor stuff is kind of "critical" for proper handling of clients. Adding other environment variables relies too much on guessing the user's intent, and the right way to do that is to use the `autostart` and `shutdown` files to set or unset additional things.

I also don't think we should be exporting a list of variables that labwc parsed into the environment. It's mostly noise, and programs that really care can just parse the environment files the same way that we do to figure out what was set.

**Comment by Vladimir-csp** [comment]
> with D-Bus, the closest that we can get is setting those variables to empty strings.

I took the same approach in uwsm, but then found out that `dbus-broker` implementation reuses systemd environment. So I've added a check to skip dbus environment management if `dbus.service` actual unit ID equals to `dbus-broker.service`.

**Comment by Vladimir-csp** [comment]
In the end, I removed any special handling of labwc environment on uwsm side (it wasn't working anyway), but decided not to set `LABWC_UPDATE_ACTIVATION_ENV=0`.
